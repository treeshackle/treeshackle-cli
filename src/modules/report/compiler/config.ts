import path from "path";
import { WORK_DIR } from "../consts";

export const WORKSPACE = path.resolve(WORK_DIR, ".treeshackle");
export const FILES_DIR_IN = path.resolve(WORKSPACE, "in");
export const FILES_DIR_OUT = path.resolve(WORKSPACE, "out");
