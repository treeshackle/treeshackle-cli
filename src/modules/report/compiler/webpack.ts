import fs from "fs";
import ufs from "unionfs";
import MemoryFs from "memory-fs";
import webpack from "webpack";
import { logger } from "../../../logger";
import { WORK_DIR } from "../consts";
import { FILES_DIR_OUT, FILES_DIR_IN } from "./config";

export const webpackConfig: (
  exclude: string[]
) => webpack.Configuration = exclude => ({
  externals: [
    (context, request, callback) => {
      logger.silly({ request, context });
      if (exclude.includes(request)) {
        callback(null, "commonjs " + request);
      } else {
        callback(null, undefined);
      }
    }
  ],
  target: "node",
  context: WORK_DIR,
  mode: "production",
  optimization: {
    runtimeChunk: {
      name: "webpack"
    }
  },
  output: {
    path: FILES_DIR_OUT
  },
  cache: false
});

export function createMemoryFs() {
  const memfs = new MemoryFs();
  logger.silly("Create in-memory folder structure");
  memfs.mkdirpSync(FILES_DIR_IN);
  memfs.mkdirpSync(FILES_DIR_OUT);
  return memfs;
}

export function createWebpackCompiler(
  memfs: MemoryFs,
  entries: webpack.Entry,
  excludedModules: string[],
  progressCallback: (progress: number) => void
) {
  const compiler = webpack({
    ...webpackConfig(excludedModules),
    entry: entries,
    plugins: [new webpack.ProgressPlugin(progressCallback)]
  });

  logger.debug("Creating union filesystem for Webpack compilation");

  compiler.inputFileSystem = ufs.use(fs).use(memfs) as any;
  compiler.outputFileSystem = memfs as any;
  return compiler;
}
