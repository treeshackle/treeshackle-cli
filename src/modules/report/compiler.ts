import fs from "fs";
import path from "path";
import ufs from "unionfs";
import MemoryFs from "memory-fs";
import webpack from "webpack";
import { logger } from "../../logger";
import { WORK_DIR } from "./consts";

const WORKSPACE = path.resolve(WORK_DIR, ".treeshackle");
const FILES_DIR_IN = path.resolve(WORKSPACE, "in");
const FILES_DIR_OUT = path.resolve(WORKSPACE, "out");

export interface ILibraryDefinition {
  name: string;
  type: string;
}

export interface ILibraryExport extends ILibraryDefinition {
  file: string;
  inpath: string;
  outpath: string;
}

export interface ISizedLibraryExport extends ILibraryExport {
  size: number;
}

// This should be customiseable
const excludeExternals = ["react", "react-dom", "styled-components"];

const webpackConfig: webpack.Configuration = {
  externals: [
    (context, request, callback) => {
      logger.silly({ request, context });
      if (excludeExternals.includes(request)) {
        callback(null, "commonjs " + request);
      } else {
        callback(null, undefined);
      }
    }
  ],
  target: "node",
  context: WORK_DIR,
  mode: "production",
  optimization: {
    runtimeChunk: {
      name: "webpack"
    }
  },
  output: {
    path: FILES_DIR_OUT
  },
  cache: false
};

logger.silly({ FILES_DIR_OUT, FILES_DIR_IN });

export async function compile(
  exportsList: ILibraryDefinition[],
  progressCallback: (progress: number) => void
): Promise<ISizedLibraryExport[]> {
  const exportDefinitions = exportsList.map(generateExportDefinition);

  logger.debug("Instansiate in-memory file system");
  const memfs = new MemoryFs();
  logger.silly("Create in-memory folder structure");
  memfs.mkdirpSync(FILES_DIR_IN);
  memfs.mkdirpSync(FILES_DIR_OUT);

  const entries = {} as webpack.Entry;
  logger.debug("Preparing to write input files to in-memory file system");

  for (let { inpath, file, name } of exportDefinitions) {
    logger.silly("Write input file to memory fs", inpath);
    logger.silly("File Content", file);
    memfs.writeFileSync(inpath, file);
    entries[name] = inpath;
  }

  logger.debug("Configuring webpack to compile in-memory files");
  const compiler = webpack({
    ...webpackConfig,
    entry: entries,
    plugins: [new webpack.ProgressPlugin(progressCallback)]
  });

  logger.debug("Creating union filesystem for Webpack compilation");
  compiler.inputFileSystem = ufs.use(fs).use(memfs) as any; // TODO: Figure out a better way to express this type
  compiler.outputFileSystem = memfs as any; // TODO: Where is the purge method on memory-fs?

  logger.debug("Initializing webpack compiler");
  try {
    const stats = await new Promise((resolve, reject) =>
      compiler.run((error, stats) => {
        logger.debug("Webpack compilation complete");
        if (error) reject(error);
        else resolve(stats);
      })
    );

    logger.silly("Webpack stats", stats);
  } catch (error) {
    logger.error("Compilation error", error);
    throw error;
  }

  const sizedExports: ISizedLibraryExport[] = [];
  logger.debug("Measuring size of compiled output files for each export");
  for (let definition of exportDefinitions) {
    try {
      logger.silly("Reading compiled file from memory fs", definition.outpath);
      const file = memfs.readFileSync(definition.outpath) as Buffer;
      //logger.silly("File content", file);
      logger.silly("Measure size of file", definition.outpath, file.byteLength);
      sizedExports.push({ ...definition, size: file.byteLength });
    } catch (error) {
      logger.error("There was an error reading the compiled resource", error);
      throw error;
    }
  }

  return sizedExports;
}

export function generateExportDefinition(
  libraryDef: ILibraryDefinition
): ILibraryExport {
  return {
    file: `import { ${libraryDef.name} as c } from "../.."; c;`,
    name: libraryDef.name,
    type: libraryDef.type,
    inpath: path.resolve(FILES_DIR_IN, `${libraryDef.name}.js`),
    outpath: path.resolve(FILES_DIR_OUT, `${libraryDef.name}.js`)
  };
}
