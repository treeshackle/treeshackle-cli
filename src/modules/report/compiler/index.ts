import path from "path";
import webpack from "webpack";
import { logger } from "../../../logger";
import { TreeshackleConfig } from "../config/schema";
import { FILES_DIR_IN, FILES_DIR_OUT } from "./config";
import { createMemoryFs, createWebpackCompiler } from "./webpack";

export interface ILibraryDefinition {
  name: string;
  type: string;
}

export interface ILibraryExport extends ILibraryDefinition {
  file: string;
  inpath: string;
  outpath: string;
}

export interface ISizedLibraryExport extends ILibraryExport {
  size: number;
}

export async function compile(
  exportsList: ILibraryDefinition[],
  config: TreeshackleConfig,
  progressCallback: (progress: number) => void
): Promise<ISizedLibraryExport[]> {
  const exportDefinitions = exportsList.map(generateExportDefinition);

  logger.debug("Instansiate in-memory file system");
  const memfs = createMemoryFs();
  const entries = {} as webpack.Entry;

  logger.debug("Preparing to write input files to in-memory file system");
  for (let { inpath, file, name } of exportDefinitions) {
    logger.debug("Write input file to memory fs", inpath);
    logger.silly("File Content", file);
    memfs.writeFileSync(inpath, file);
    entries[name] = inpath;
  }

  logger.debug("Configuring webpack to compile in-memory files");
  const compiler = createWebpackCompiler(
    memfs,
    entries,
    config.modules.exclude,
    progressCallback
  );

  logger.debug("Initializing webpack compiler");
  try {
    const stats = await new Promise((resolve, reject) =>
      compiler.run((error, stats) => {
        logger.debug("Webpack compilation complete");
        if (error) reject(error);
        else resolve(stats);
      })
    );

    logger.silly("Webpack stats", stats);
  } catch (error) {
    logger.error("Compilation error", error);
    throw error;
  }

  const sizedExports: ISizedLibraryExport[] = [];
  logger.debug("Measuring size of compiled output files for each export");
  for (let definition of exportDefinitions) {
    try {
      logger.silly("Reading compiled file from memory fs", definition.outpath);
      const file = memfs.readFileSync(definition.outpath) as Buffer;
      logger.silly("File content", file);
      logger.silly("Measure size of file", definition.outpath, file.byteLength);
      sizedExports.push({ ...definition, size: file.byteLength });
    } catch (error) {
      logger.error("There was an error reading the compiled resource", error);
      throw error;
    }
  }

  return sizedExports;
}

export function generateExportDefinition(
  libraryDef: ILibraryDefinition
): ILibraryExport {
  return {
    file: `import { ${libraryDef.name} as c } from "../.."; c;`,
    name: libraryDef.name,
    type: libraryDef.type,
    inpath: path.resolve(FILES_DIR_IN, `${libraryDef.name}.js`),
    outpath: path.resolve(FILES_DIR_OUT, `${libraryDef.name}.js`)
  };
}
