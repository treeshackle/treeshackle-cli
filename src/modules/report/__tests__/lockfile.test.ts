import { ISizedLibraryExport } from "../compiler";
import {
  generateLockfile,
  getExportType,
  createReport,
  diffReport,
  IExportReport
} from "../lockfile";

const libExports: ISizedLibraryExport[] = [
  {
    name: "foo",
    size: 12345,
    file: "foo/bar.js",
    inpath: "in/path.js",
    outpath: "out/path.js",
    type: "function"
  },
  {
    name: "bar",
    size: 987,
    file: "bar/foo.js",
    inpath: "path/in.js",
    outpath: "path/out.js",
    type: "function"
  }
];

const testReport1: IExportReport = {
  fooBar: {
    size: "1.81 kB",
    type: "function"
  },
  bazBar: {
    size: "5.81 kB",
    type: "function"
  },
  test: {
    size: "12.3 kB",
    type: "function"
  }
};

const testReport2: IExportReport = {
  fooBar: {
    size: "2.81 kB",
    type: "function"
  },
  bazBar: {
    size: "5.81 kB",
    type: "string"
  },
  bangBox: {
    size: "480 B",
    type: "string"
  }
};

describe("generateLockfile", () => {
  it("should match the lockfile snapshot", () => {
    expect(generateLockfile(testReport1)).toMatchSnapshot();
  });
});

describe("getExportType", () => {
  it("should return the type of the export", () => {
    // Initial version of treeshackle is only going to be able to
    // get the primitive type.
    expect(getExportType("foo")).toBe("string");
    expect(getExportType(() => {})).toBe("function");
    expect(getExportType(4)).toBe("number");
    expect(getExportType({})).toBe("object");
  });
});

describe("createReport", () => {
  expect(createReport(libExports)).toMatchSnapshot();
});

describe("diffReport", () => {
  it("should match the diff snapshot", () => {
    expect(diffReport(testReport1, testReport2)).toMatchSnapshot();
  });
});
