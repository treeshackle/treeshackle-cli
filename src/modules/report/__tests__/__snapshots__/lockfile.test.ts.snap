// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "bar": Object {
    "size": "987 B",
    "type": "function",
  },
  "foo": Object {
    "size": "12.3 kB",
    "type": "function",
  },
}
`;

exports[`diffReport should match the diff snapshot 1`] = `
Array [
  Object {
    "after": Object {
      "size": "2.81 kB",
      "type": "function",
    },
    "before": Object {
      "size": "1.81 kB",
      "type": "function",
    },
    "name": "fooBar",
    "type": "change",
  },
  Object {
    "after": Object {
      "size": "5.81 kB",
      "type": "string",
    },
    "before": Object {
      "size": "5.81 kB",
      "type": "function",
    },
    "name": "bazBar",
    "type": "change",
  },
  Object {
    "export": Object {
      "size": "12.3 kB",
      "type": "function",
    },
    "name": "test",
    "type": "removal",
  },
  Object {
    "export": Object {
      "size": "480 B",
      "type": "string",
    },
    "name": "bangBox",
    "type": "addition",
  },
]
`;

exports[`generateLockfile should match the lockfile snapshot 1`] = `
"{
  \\"treeshackle\\": {
    \\"version\\": \\"1.0.0-test\\"
  },
  \\"exports\\": {
    \\"fooBar\\": {
      \\"size\\": \\"1.81 kB\\",
      \\"type\\": \\"function\\"
    },
    \\"bazBar\\": {
      \\"size\\": \\"5.81 kB\\",
      \\"type\\": \\"function\\"
    },
    \\"test\\": {
      \\"size\\": \\"12.3 kB\\",
      \\"type\\": \\"function\\"
    }
  }
}"
`;
